File: Day4/main.c
#include <stdio.h>
#include <string.h>

#if 0
#define FILENAME "sample_data.txt"
#else
#define FILENAME "data.txt"
#endif

int main()
{
    FILE *fp = fopen(FILENAME, "r");
    if (fp == NULL)
    {
        printf("Error opening file " FILENAME "\n");
        return 1;
    }
    char buf[256];
    int count_part1 = 0;
    int count_part2 = 0;
    while (fgets(buf, sizeof(buf), fp))
    {
        int x1, y1, x2, y2;
        sscanf(buf, "%d-%d,%d-%d", &x1, &y1, &x2, &y2);
        if (((y1 <= y2) && (x1 >= x2)) || ((y1 >= y2) && (x1 <= x2)))
        {
            count_part1 += 1;
            count_part2 += 1;
        }
        else if (((x1 >= x2) && (x1 <= y2)) || ((x2 >= x1) && (x2 <= y1)))
        {
            count_part2 += 1;
        }
    }
    printf("Part 1: %d\nPart 2: %d\n", count_part1, count_part2);
    return 0;
}

File: Day4/main.py
import re
SAMPLE_DATA: bool = False
filename = "sample_data.txt" if SAMPLE_DATA else "data.txt"
with open(filename) as f:
    count_part1 = 0
    count_part2 = 0
    [(count_part1 := count_part1 + (((y1 <= y2) and (x1 >= x2)) or ((y1 >= y2) and (x1 <= x2))), count_part2 := count_part2 + (((x1 >= x2) and (x1 <= y2)) or ((x2 >= x1) and (x2 <= y1)))) for x1, y1, x2, y2 in ((int(x) for x in re.split(r"[-,]", line)) for line in f.readlines())]
print(f'{count_part1} {count_part2}')